Trusted Types does excess work when not set in CSP

lukewarlow opened the pull request:
#### 160b58921311c5f62fe63c7b32d2b7a9e184f1ca
<pre>
Trusted Types does excess work when not set in CSP
<a href="https://bugs.webkit.org/show_bug.cgi?id=287107">https://bugs.webkit.org/show_bug.cgi?id=287107</a>

Reviewed by Ryosuke Niwa.

This patch improves the CSP check within trustedTypeCompliantString,
it now uses a flag on the document rather than looping through all policies.

Attribute functions are also updated to skip work not needed when
trusted types isn&apos;t required by CSP.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setRequiresTrustedTypes):
* Source/WebCore/dom/Document.h:
(WebCore::Document::requiresTrustedTypes const):
* Source/WebCore/dom/Element.cpp:
(WebCore::trustedTypesCompliantAttributeValue):
(WebCore::Element::setAttribute):
(WebCore::Element::setAttributeNode):
(WebCore::Element::setAttributeNodeNS):
(WebCore::Element::setAttributeNS):
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::trustedTypeCompliantString):

Canonical link: <a href="https://commits.webkit.org/290561@main">https://commits.webkit.org/290561@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd77c2b195a98a966516f4485d57811155a56f95

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90358 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9888 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45284 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95363 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41134 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92410 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10275 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18205 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69557 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27137 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93359 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7870 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81968 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49918 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7598 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40265 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77928 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37401 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97190 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17546 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12906 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78550 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17803 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77796 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77771 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22220 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20830 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10811 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14230 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17556 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22883 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17297 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20749 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19081 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4485,6 +4485,8 @@ void Document::setRequiresTrustedTypes(bool required)
         return;
 
     frame->checkedScript()->setRequiresTrustedTypes(required);
```
Do we need to call this when the pref is disabled?

lukewarlow replied:
I think yes, in case of weird situations where the JSC flag is enabled but this one isn't you could get into odd situations otherwise.

nt1m replied:
I don't think we need to handle strange combinations like that. Maybe add an assertion to check they both use the same value.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4485,6 +4485,8 @@ void Document::setRequiresTrustedTypes(bool required)
         return;
 
     frame->checkedScript()->setRequiresTrustedTypes(required);
+    if (scriptExecutionContext()->settingsValues().trustedTypesEnabled)
```
Can you check `settings().trustedTypesEnabled()` instead?

lukewarlow replied:
Ah yeah good point. Will change this and Document write to use that instead.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2068,13 +2067,17 @@ ExceptionOr<void> Element::setAttribute(const AtomString& qualifiedName, const T
     else {
-        auto attributeTypeAndSink = trustedTypeForAttribute(nodeName(), name.localName().convertToASCIILowercase(), this->namespaceURI(), name.namespaceURI());
+        AttributeTypeAndSink attributeTypeAndSink;
+        if (document().requiresTrustedTypes())
+            attributeTypeAndSink = trustedTypeForAttribute(nodeName(), name.localName().convertToASCIILowercase(), this->namespaceURI(), name.namespaceURI());
         auto attributeValue = trustedTypesCompliantAttributeValue(attributeTypeAndSink.attributeType, value, this, attributeTypeAndSink.sink);
```
Do we need to invoke this when trusted types are not used? I would just change `if (!document().scriptExecutionContext()->settingsValues().trustedTypesEnabled)` to `if (!document().requiresTrustedTypes())` above if not.

lukewarlow replied:
Yes because we need to extract the string from the TrustedType objects passed through to the function.

Even if TT isn't enforced in CSP they can still create the objects so we need to support that scenario.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3646,7 +3649,9 @@ ExceptionOr<void> Element::setAttributeNS(const AtomString& namespaceURI, const
         QualifiedName parsedAttributeName  = result.returnValue();
-        auto attributeTypeAndSink = trustedTypeForAttribute(nodeName(), parsedAttributeName.localName(), this->namespaceURI(), parsedAttributeName.namespaceURI());
+        AttributeTypeAndSink attributeTypeAndSink;
+        if (document().requiresTrustedTypes())
```
Same question here

lukewarlow replied:
See above.

nt1m commented on Source/WebCore/dom/TrustedType.cpp:
```diff
@@ -163,9 +163,15 @@ ExceptionOr<String> trustedTypeCompliantString(TrustedType expectedType, ScriptE
 
-    CheckedPtr contentSecurityPolicy = scriptExecutionContext.checkedContentSecurityPolicy();
+    auto requireTrustedTypes = false;
 
-    auto requireTrustedTypes = contentSecurityPolicy && contentSecurityPolicy->requireTrustedTypesForSinkGroup("script"_s);
+    if (RefPtr document = dynamicDowncast<Document>(scriptExecutionContext))
+        requireTrustedTypes = document->requiresTrustedTypes();
+    else {
+        CheckedPtr contentSecurityPolicy = scriptExecutionContext.checkedContentSecurityPolicy();
+
+        requireTrustedTypes = contentSecurityPolicy && contentSecurityPolicy->requireTrustedTypesForSinkGroup("script"_s);
+    }
```
I'm not sure I understand this change

lukewarlow replied:
It uses the document's flag when it can, but for other envs (such as Workers) we fallback to the old method of checking the CSP.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2068,13 +2067,17 @@ ExceptionOr<void> Element::setAttribute(const AtomString& qualifiedName, const T
     if (!document().scriptExecutionContext()->settingsValues().trustedTypesEnabled)
```
Can you change this to use `settings().trustedTypesEnabled()` ?

nt1m replied:
ditto in `Element::setAttributeNS` (and other `Element` call sites)

lukewarlow committed:
Trusted Types does excess work when not set in CSP
https://bugs.webkit.org/show_bug.cgi?id=287107

Reviewed by Ryosuke Niwa.

This patch improves the CSP check within trustedTypeCompliantString,
it now uses a flag on the document rather than looping through all policies.

Attribute functions are also updated to skip work not needed when
trusted types isn't required by CSP.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setRequiresTrustedTypes):
* Source/WebCore/dom/Document.h:
(WebCore::Document::requiresTrustedTypes const):
* Source/WebCore/dom/Element.cpp:
(WebCore::trustedTypesCompliantAttributeValue):
(WebCore::Element::setAttribute):
(WebCore::Element::setAttributeNode):
(WebCore::Element::setAttributeNodeNS):
(WebCore::Element::setAttributeNS):
* Source/WebCore/dom/TrustedType.cpp:
(WebCore::trustedTypeCompliantString):

Canonical link: https://commits.webkit.org/290561@main
